src/main/java/jenkins/plugins/git/GitSCMSource.java

```diff
@@ -123,21 +127,59 @@ public String getDisplayName() {
         }
 
         public ListBoxModel doFillCredentialsIdItems(@AncestorInPath SCMSourceOwner context,
-                                                     @QueryParameter String remote) {
+                                                     @QueryParameter String remote,
+                                                     @QueryParameter String credentialsId) {
             if (context == null || !context.hasPermission(Item.CONFIGURE)) {
-                return new ListBoxModel();
+                return new StandardListBoxModel().includeCurrentValue(credentialsId);
             }
-            StandardListBoxModel result = new StandardListBoxModel();
-            result.withEmptySelection();
-            result.withMatching(GitClient.CREDENTIALS_MATCHER,
-                    CredentialsProvider.lookupCredentials(
-                            StandardUsernameCredentials.class,
+            return new StandardListBoxModel()
+                    .includeEmptyValue()
+                    .includeMatchingAs(
+                            context instanceof Queue.Task ? Tasks.getAuthenticationOf((Queue.Task)context) : ACL.SYSTEM,
                             context,
-                            ACL.SYSTEM,
-                            URIRequirementBuilder.fromUri(remote).build()
-                    )
-            );
-            return result;
+                            StandardUsernameCredentials.class,
+                            URIRequirementBuilder.fromUri(remote).build(),
+                            GitClient.CREDENTIALS_MATCHER)
+                    .includeCurrentValue(credentialsId);
+        }
+
+        public FormValidation doCheckCredentialsId(@AncestorInPath SCMSourceOwner context,
+                                                   @QueryParameter String url,
+                                                   @QueryParameter String value) {
+            if (context == null || !context.hasPermission(Item.CONFIGURE)) {
+                return FormValidation.ok();
+            }
+
+            value = Util.fixEmptyAndTrim(value);
+            if (value == null) {
+                return FormValidation.ok();
+            }
+
+            url = Util.fixEmptyAndTrim(url);
+            if (url == null)
+            // not set, can't check
+            {
+                return FormValidation.ok();
+            }
+
+            for (ListBoxModel.Option o : CredentialsProvider.listCredentials(
+                    StandardUsernameCredentials.class,
+                    context,
+                    context instanceof Queue.Task
```

Comment:
it could as `public interface SCMSourceOwner extends Item;` so there is nothing stopping an implementation also being a `Queue.Task` (which might make sense if you want the branch indexing to actually be listed on the queue)